android/PhysicalWeb/app/src/main/java/org/physical_web/physicalweb/PhysicalWebBroadcastService.java

```diff
@@ -98,40 +102,58 @@ public void onCreate() {
-            // Set the url if we can
-            Log.d(TAG, "valid length");
-            mShareUrl = mDisplayUrl;
-            broadcastUrl();
+        mStartedByRestart = (intent == null);
+        fetchBroadcastData(intent);
+        if (mDisplayUrl != null){
+            Log.d(TAG, "SERVICE onStartCommand");
+            IntentFilter filter = new IntentFilter(BluetoothAdapter.ACTION_STATE_CHANGED);
+            registerReceiver(mReceiver, filter);
+            
+            Log.d(TAG, mDisplayUrl);
+
+            if (hasValidUrlLength(mDisplayUrl) && checkAndHandleAsciiUrl(mDisplayUrl)) {
+                // Set the url if we can
+                Log.d(TAG, "valid length");
+                mShareUrl = mDisplayUrl;
+                broadcastUrl();
+            } else {
+                Log.d(TAG, "needs shortening");
+                // Shorten the url if necessary
+                UrlShortenerClient.ShortenUrlCallback urlSetter =
+                new UrlShortenerClient.ShortenUrlCallback() {
+                    @Override
+                    public void onUrlShortened(String newUrl) {
+                        Log.d(TAG, "shortening success");
+                        mShareUrl = newUrl;
+                        broadcastUrl();
+                    }
+                    @Override
+                    public void onError(String oldUrl) {
+                        Toast.makeText(getApplicationContext(), getString(R.string.shorten_error),
+                            Toast.LENGTH_LONG).show();
+                    }
+                };
+                UrlShortenerClient.getInstance(this).shortenUrl(mDisplayUrl, urlSetter, TAG);
+            }
         } else {
-            Log.d(TAG, "needs shortening");
-            // Shorten the url if necessary
-            UrlShortenerClient.ShortenUrlCallback urlSetter =
-            new UrlShortenerClient.ShortenUrlCallback() {
-                @Override
-                public void onUrlShortened(String newUrl) {
-                    Log.d(TAG, "shortening success");
-                    mShareUrl = newUrl;
-                    broadcastUrl();
-                }
-                @Override
-                public void onError(String oldUrl) {
-                    Toast.makeText(getApplicationContext(), getString(R.string.shorten_error),
-                        Toast.LENGTH_LONG).show();
-                }
-            };
-            UrlShortenerClient.getInstance(this).shortenUrl(mDisplayUrl, urlSetter, TAG);
+            onDestroy();
         }
-        // Keep the service running
         return START_STICKY;
     }
 
+    private void fetchBroadcastData(Intent intent){
+        if (intent == null){
+            SharedPreferences sharedPrefs = PreferenceManager.getDefaultSharedPreferences(this);
+            mDisplayUrl = sharedPrefs.getString(PREVIOUS_BROADCAST_URL_KEY, null);
+        } else{
+            SharedPreferences sharedPrefs = PreferenceManager.getDefaultSharedPreferences(this);
```

Comment:
I prefer chaining, i.e.

    PreferenceManager.getDefaultSharedPreferences(this).edit()
        .putString(PREVIOUS_BROADCAST_URL_KEY, mDisplayUrl)
        .commit();